### PR TITLE
GH-42139: [C++] Fix some potential uninitialized variable warnings

### PR DESCRIPTION
**Pull request by raulcd**
### Rationale for this change

Fix a couple warnings about potentially uninitialized variables

### What changes are included in this PR?

Fix the warnings

### Are these changes tested?

Test via CI.

### Are there any user-facing changes?

No
* GitHub Issue: #42139

### CODE REVIEW

**Review by felipecrv — APPROVED** [review]

**Comment by raulcd** [comment]
thanks for approving @felipecrv 

Do you know how I could reproduce the other error raised on the issue? I can't seem to be able:
```
[265/572] Building CXX object src/arrow/CMakeFiles/arrow_objlib.dir/c/bridge.cc.o
/home/antoine/arrow/dev/cpp/src/arrow/c/bridge.cc: In member function 'arrow::Status arrow::{anonymous}::ArrayExporter::Export(const std::shared_ptr<arrow::ArrayData>&)':
/home/antoine/arrow/dev/cpp/src/arrow/c/bridge.cc:540:33: warning: '<unnamed>.arrow::<unnamed>::ArrayExporter::export_.arrow::{anonymous}::ExportedArrayPrivateData::dictionary_' may be used uninitialized [-Wmaybe-uninitialized]
  540 |   ARROW_DEFAULT_MOVE_AND_ASSIGN(ExportedArrayPrivateData);
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~~
/home/antoine/arrow/dev/cpp/src/arrow/c/bridge.cc:620:68: note: '<anonymous>' declared here
  620 |       child_exporters_.emplace_back(ArrayExporter{device_interface_});
      |                                                                    ^
```
See my comment on: https://github.com/apache/arrow/issues/42139#issuecomment-2176586648

**Comment by felipecrv** [comment]
@raulcd I can repro this with GCC. The fix is simple:

```diff
diff --git a/cpp/src/arrow/c/bridge.cc b/cpp/src/arrow/c/bridge.cc
index afb664c3b..56f3de473 100644
--- a/cpp/src/arrow/c/bridge.cc
+++ b/cpp/src/arrow/c/bridge.cc
@@ -528,7 +528,7 @@ namespace {
 struct ExportedArrayPrivateData : PoolAllocationMixin<ExportedArrayPrivateData> {
   // The buffers are owned by the ArrayData member
   SmallVector<const void*, 3> buffers_;
-  struct ArrowArray dictionary_;
+  struct ArrowArray dictionary_{};
   SmallVector<struct ArrowArray, 1> children_;
   SmallVector<struct ArrowArray*, 4> child_pointers_;
```

File: cpp/src/arrow/c/bridge.cc [hunk 1]
```diff
@@ -528,7 +528,7 @@ namespace {
 struct ExportedArrayPrivateData : PoolAllocationMixin<ExportedArrayPrivateData> {
   // The buffers are owned by the ArrayData member
   SmallVector<const void*, 3> buffers_;
-  struct ArrowArray dictionary_;
+  struct ArrowArray dictionary_{};
```
Comment (felipecrv) [inline cpp/src/arrow/c/bridge.cc:531]:
```suggestion
  struct ArrowArray dictionary_ {};
```

**Comment by raulcd** [comment]
CI failures seem unrelated

**Review by pitrou — APPROVED** [review]
This fixes the problem for me. Thanks @raulcd !

**Pull request merged**